File: deck/seed/code/native/rust/runtime/pipe.rs
// Pipe runtime: stream one child's stdout into another child's stdin until the source closes.
// Each dock handle carries the platform process as Rc<RefCell<std::process::Child>>; a handle of
// the wrong shape is a no-op. Reached only through the public process API.
mod pipe_runtime {
    use std::io::{Read, Write};

    fn of(dock: &std::rc::Rc<dyn std::any::Any>) -> Option<&std::cell::RefCell<std::process::Child>> {
        dock.downcast_ref::<std::cell::RefCell<std::process::Child>>()
    }

    pub async fn connect(from: std::rc::Rc<dyn std::any::Any>, to: std::rc::Rc<dyn std::any::Any>) {
        let stdout = of(&from).and_then(|cell| cell.borrow_mut().stdout.take());
        let stdin = of(&to).and_then(|cell| cell.borrow_mut().stdin.take());
        if let (Some(mut stdout), Some(mut stdin)) = (stdout, stdin) {
            let mut buf = [0u8; 8192];
            loop {
                let n = match stdout.read(&mut buf) {
                    Ok(0) | Err(_) => break,
                    Ok(n) => n,
                };
                if stdin.write_all(&buf[..n]).is_err() {
                    break;
                }
            }
        }
    }
}

File: deck/seed/code/native/kotlin/runtime/pipe.kt
// Pipe runtime for the kotlin target: stream one child's standard output into another child's standard input
// until the source closes. Each dock handle carries the platform process as a java.lang.Process; a handle of the
// wrong shape is a no-op. Reached only through the public process API.
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext

object pipeRuntime {
  private fun of(dock: Any?): Process? = dock as? Process

  suspend fun connect(from: Any?, to: Any?) {
    withContext(Dispatchers.IO) {
      val out = of(from)?.inputStream ?: return@withContext
      val into = of(to)?.outputStream ?: return@withContext
      val buffer = ByteArray(8192)

      try {
        while (true) {
          val read = out.read(buffer)

          if (read <= 0) break

          into.write(buffer, 0, read)
          into.flush()
        }
      } catch (error: Exception) {
        return@withContext
      } finally {
        try {
          into.close()
        } catch (error: Exception) {
          Unit
        }
      }
    }
  }
}
